game/src/sandbox/dashboards/trip_table.rs

```
@@ -4,52 +4,177 @@ use abstutil::prettyprint_usize;
+        .build(ctx);
+
+        Self {
+            tab: DashTab::TripTable,
+            table_tabs: tabs,
+            panel,
+            finished_trips_table,
+            cancelled_trips_table,
+            unfinished_trips_table,
+        }
     }
 }
 
-pub struct UnfinishedTripTable;
-
-impl UnfinishedTripTable {
-    pub fn new(ctx: &mut EventCtx, app: &App) -> Box<dyn State<App>> {
-        GenericTripTable::new(
-            ctx,
-            app,
-            DashTab::UnfinishedTripTable,
-            make_table_unfinished_trips(app),
-            make_panel_unfinished_trips,
-        )
+impl State<App> for TripTable {
+    fn event(&mut self, ctx: &mut EventCtx, app: &mut App) -> Transition {
+        match self.panel.event(ctx) {
+            Outcome::Clicked(x) => {
+                if self.table_tabs.active_tab_idx() == 0 && self.finished_trips_table.clicked(&x) {
+                    self.finished_trips_table
+                        .replace_render(ctx, app, &mut self.panel);
+                } else if self.table_tabs.active_tab_idx() == 1
+                    && self.cancelled_trips_table.clicked(&x)
+                {
+                    self.cancelled_trips_table
+                        .replace_render(ctx, app, &mut self.panel);
+                } else if self.table_tabs.active_tab_idx() == 2
+                    && self.unfinished_trips_table.clicked(&x)
+                {
+                    self.unfinished_trips_table
+                        .replace_render(ctx, app, &mut self.panel);
+                } else if let Ok(idx) = x.parse::<usize>() {
+                    return open_trip_transition(app, idx);
+                } else if x == "close" {
+                    return Transition::Pop;
+                } else if self.table_tabs.handle_action(ctx, &x, &mut self.panel) {
+                    // if true, tabs handled the action
+                } else {
+                    todo!("unhandled action: {}", x)
```

nit: `panic` instead of `todo`, since this should logically be unreachable in its current form, right?

---

done!